File: test_modern_perl.pl
use feature 'try';
try {
    dangerous_code();
} catch ($e) {
    warn "Error: $e";
} finally {
    cleanup();
}

use feature 'defer';
{
    defer { print "cleanup" }
    do_work();
}

use feature 'class';
class Point {
    field $x :param = 0;
    field $y :param = 0;
    
    method move ($dx, $dy) {
        $x += $dx;
        $y += $dy;
    }
}

File: crates/perl-parser/examples/test_modern_perl.rs
//! Test modern Perl features
use perl_parser::Parser;

fn main() {
    let tests = vec![
        // Version declaration
        ("use v5.36;", "version declaration"),
        ("use v5.32.1;", "version with patch level"),
        ("use 5.036;", "numeric version"),
        // Try/catch
        ("try { die 'oops' } catch ($e) { warn $e }", "try/catch"),
        ("try { risky() } catch ($e) { handle($e) }", "try/catch with calls"),
        // Defer
        ("defer { cleanup() }", "defer block"),
        // Class/method (Corinna)
        ("class Point { }", "empty class"),
        ("class Point { field $x; }", "class with field"),
        ("class Point { method new() { } }", "class with method"),
        ("method new($x, $y) { }", "method with signature"),
        // Field declarations
        ("field $x;", "field declaration"),
        ("field $x = 42;", "field with default"),
    ];

    for (test, desc) in tests {
        println!("\n=== Testing: {} ===", desc);
        println!("Input: {}", test);
        let mut parser = Parser::new(test);
        match parser.parse() {
            Ok(ast) => {
                println!("✅ Success! S-expr: {}", ast.to_sexp());
            }
            Err(e) => {
                println!("❌ Error: {}", e);
            }
        }
    }
}
